pkg/controller/operators/catalog/operator.go

```diff
@@ -569,3 +454,340 @@ func (o *Operator) ExecutePlan(plan *v1alpha1.InstallPlan) error {
 
 	return nil
 }
+
+// DependencyResolver defines how a something that resolves dependencies (CSVs, CRDs, etc...)
+// should behave
+type DependencyResolver interface {
+	ResolveInstallPlan(firstSrcKey sourceKey, sources map[sourceKey]registry.Source, plan *v1alpha1.InstallPlan) error
+	ResolveCSV(csvName string, firstSrcKey sourceKey, sources map[sourceKey]registry.Source) (stepResourceMap, error)
+	ResolveCRDDescription(crdDesc csvv1alpha1.CRDDescription, firstSrcKey sourceKey, sources map[sourceKey]registry.Source, owned bool) (v1alpha1.StepResource, string, error)
+}
+
+// SingleSourceResolver resolves dependencies from a single CatalogSource
+type SingleSourceResolver struct{}
+
+// ResolveInstallPlan resolves all dependencies for an InstallPlan
+func (resolver *SingleSourceResolver) ResolveInstallPlan(firstSrcKey sourceKey, sources map[sourceKey]registry.Source, plan *v1alpha1.InstallPlan) error {
+	srm := make(stepResourceMap)
+	for _, csvName := range plan.Spec.ClusterServiceVersionNames {
+		csvSRM, err := resolver.ResolveCSV(csvName, firstSrcKey, sources)
+		if err != nil {
+			return err
+		}
+
+		srm.Combine(csvSRM)
+	}
+
+	plan.Status.Plan = srm.Plan()
```

Comment:
Eww side effects

Reply:
This was like this originally - Would you prefer a return value and the plan field being set by the caller?